File: 238.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        def get_mul(_nums):
            _res = 1
            for _num in _nums:
                _res *= _num
            return _res
        res, cache = [], {}
        for i, num in enumerate(nums):
            if num not in cache:
                cache[num] = get_mul(nums[:i] + nums[i+1:])
            res.append(cache[num])
        return res

File: 238.go
package leetcode

import "fmt"

func main() {
	params := []int{1, 2,3,4}
	productExceptSelf(params)
}

func productExceptSelf(nums []int) []int {
	var res []int
	cache := make(map[int]int)
	for i, num := range nums {
		_, found := cache[num]
		if !found {
			cache[num] = getMul(append(nums[:i], nums[i+1:len(nums)]...))
		}
		res = append(res, cache[num])
	}
	return res
}

func getMul(nums []int) int {
	res := 1
	for _, num := range nums {
		res *= num
	}
	fmt.Printf("%d", res)
	return res
}